### PR TITLE
Fix to route a link to the specified page with client-side rendering

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

Path patterns are used folder structures in Next.js such as '/app/projects/[projectNames]'. If '/app/projects/myProj' is requested to Central Dogma server, the server fails to find
'app/projects/myProj/index.html' file and returns 404 not found. This does not happen if a client app starts with `npm run xxx` commands because the links are handled by Next.js server and only REST APIs are requested to a Central Dogma server.

Modifications:

- Return `index.html` file as a fallback if the `FileService` for new webapp fails to find a file.
- Trigger Next.js router to route to the desired page when a page is loaded for the first time.

Result:

You can now visit a Central Dogma webapp with a deep link.

### CODE REVIEW

**Comment by ikhoon** [comment]
@rainy789 Could you take a look if available? :wink:

File: webapp/src/pages/_app.tsx [hunk 1]
```diff
@@ -20,8 +20,18 @@ type AppPropsWithLayout = AppProps & {
...
 
-const DogmaApp = ({ Component, pageProps }: AppPropsWithLayout) => {
+const DogmaApp = ({Component, pageProps}: AppPropsWithLayout) => {
   const router = useRouter();
+  useEffect(() => {
+    // Path patterns are used folder structures in Next.js such as '/app/projects/[projectNames]'.
+    // If '/app/projects/myProj' is requested to Central Dogma server, the server fails to find
+    // 'app/projects/myProj/index.html' file and returns 'index.html' as a fallback and the landing page is
+    // rendered instead. As a workaround, this triggers Next.js router to route to the desired page when a page
+    // is loaded for the first time.
```
Comment (jrhee17) [inline webapp/src/pages/_app.tsx:30]: > is loaded for the first time.

It seems like `router.asPath` is always pushed every time a state change is made. I'm wondering if this is the intended behavior?

**Review by jrhee17 — APPROVED** [review]
I don't want to be a blocker so approving 😄 

Also, when I tried running the webapp I got stuck at the following page. Any pointers would be helpful 🙏 
Screen Shot 2023-03-03 at 1 10 27 PM

File: server/src/main/java/com/linecorp/centraldogma/server/CentralDogma.java [hunk 2]
```diff
@@ -744,11 +744,15 @@ public String serviceName(ServiceRequestContext ctx) {
...
+            final HttpService fallbackFile = HttpFile.of(CentralDogma.class.getClassLoader(),
+                                                         "com/linecorp/centraldogma/webapp/index.html")
+                                                     .asService();
             sb.serviceUnder("/",
                             FileService.builder(CentralDogma.class.getClassLoader(),
                                                 "com/linecorp/centraldogma/webapp")
                                        .cacheControl(ServerCacheControl.REVALIDATED)
-                                       .build());
+                                       .build()
+                                       .orElse(fallbackFile));
```
[the remark this answers is not in the document]
Comment (ikhoon) [reply server/src/main/java/com/linecorp/centraldogma/server/CentralDogma.java:755]: My explanation was insufficient. The actual HTML file path is `/app/projects/[projectName]/index.html` but the request path for the file is `/app/projects/myProj`.
FileService tries to find a matched file in `/app/projects/myProj` and `/app/projects/myProj/index.html`. So no page is rendered.

> Question) I think sometimes returning a 404 makes sense. 

There is a 404 page on the client-side. Next.js router may redirect to 404 if there are no matched routes.

File: webapp/src/pages/_app.tsx (hunk 1 above)
Comment (ikhoon) [reply webapp/src/pages/_app.tsx:30]: This should be called only once when the web application is initialized.
Is this repeatedly called?

File: server/src/main/java/com/linecorp/centraldogma/server/CentralDogma.java [hunk 3]
```diff
@@ -744,11 +744,15 @@ public String serviceName(ServiceRequestContext ctx) {
                 sb.service(LOGOUT_PATH, authProvider.webLogoutService());
             }
 
+            final HttpService fallbackFile = HttpFile.of(CentralDogma.class.getClassLoader(),
```
Comment (trustin) [inline server/src/main/java/com/linecorp/centraldogma/server/CentralDogma.java:747]: `fallbackFileService`?

File: webapp/src/pages/_app.tsx [hunk 4]
```diff
@@ -20,8 +20,18 @@ type AppPropsWithLayout = AppProps & {
   Component: NextPageWithLayout;
 };
 
-const DogmaApp = ({ Component, pageProps }: AppPropsWithLayout) => {
+const DogmaApp = ({Component, pageProps}: AppPropsWithLayout) => {
   const router = useRouter();
+  useEffect(() => {
+    // Path patterns are used folder structures in Next.js such as '/app/projects/[projectNames]'.
```
Comment (trustin) [inline webapp/src/pages/_app.tsx:26]: Maybe rephrase?

**Comment by ikhoon** [comment]
> Also, when I tried running the server with the built webapp I got stuck at the following page. Any pointers would be helpful 🙏

How did you start the test server?

**Comment by ikhoon** [comment]
I found a bug that `secury_enabled` is handled by the file service.
We may need to set `app` as a context path for the file service.

**Comment by ikhoon** [comment]
The bug has been fixed by [`90ce304` (#814)](https://github.com/line/centraldogma/pull/814/commits/90ce30461bcf9ca70007f5ef6b63afbf6abc1fa9)

**Review by trustin — APPROVED** [review]

File: webapp/src/pages/_app.tsx [hunk 5]
```diff
@@ -22,6 +22,18 @@ type AppPropsWithLayout = AppProps & {
...
+  useEffect(() => {
+    // Next.js uses a path pattern to dynamically match the request path to a specific HTML file.
+    // For example, `/app/projects/[projectNames]/index.html' is generated to render `/app/projects/myProj`
+    // page. `FileService` serving the static resources for Central Dogma webapp does not understand the
+    // path pattern syntax in the folder names. If `/ap/projects/myProj` is requested, the FileService will try
+    // to find '/app/projects/myProj/index.html' as is, and 'index.html' is returned as a fallback. As a
+    // workaround, this triggers Next.js router to route to the desired page when a page is loaded for the
+    // first time.
+    if (router.asPath !== '/') {
+      Router.push(router.asPath);
```
Comment (rainy789) [inline webapp/src/pages/_app.tsx:34]:
```suggestion
      router.push(router.asPath);
```

File: webapp/src/pages/_app.tsx [hunk 6]
```diff
@@ -4,8 +4,8 @@ import { Provider } from 'react-redux';
 import { ChakraProvider, theme } from '@chakra-ui/react';
 import { Authorized } from 'dogma/features/auth/Authorized';
 import { NextPage } from 'next';
-import { ReactElement, ReactNode } from 'react';
-import { useRouter } from 'next/router';
+import { ReactElement, ReactNode, useEffect } from 'react';
+import Router, { useRouter } from 'next/router';
 import { Layout } from 'dogma/common/components/Layout';
```
Comment (rainy789) [inline webapp/src/pages/_app.tsx:9]: We can also get `router` from `router = useRouter()`, they are actually similar. I think the difference is the non global `router` will return a different path when we call `router.asPath`.

**Comment by rainy789** [comment]
> * Return `index.html` file as a fallback if the `FileService` for new webapp fails to find a file.

Question: can this be fixed by adding a trailing slash? https://nextjs.org/docs/api-reference/next.config.js/exportPathMap#adding-a-trailing-slash

File: webapp/src/pages/_app.tsx (hunk 1 above)
Comment (jrhee17) [reply webapp/src/pages/_app.tsx]: I'm not sure, but aren't effects called every time a state is changed?

Comment (ikhoon) [reply webapp/src/pages/_app.tsx]: `[]` is set to the dependency of `useEffect()` so it should be run once.
https://css-tricks.com/run-useeffect-only-once/#aa-the-trick-is-that-useeffect-takes-a-second-parameter

File: webapp/src/pages/_app.tsx (hunk 5 above)
Comment (ikhoon) [reply webapp/src/pages/_app.tsx:34]: `replace` looks better. Let me try to use it.

**Comment by ikhoon** [comment]
It may be not related to a trailing slash. As you know, the HTML file for a project is stored at `/app/products/[projectName]/index.html`. If I directly type, click or refresh `https://mydogma.apps.com/app/products/my-project`, a Next.js server interprets the URL and then serves `/app/products/[projectName]/index.html` correctly from the file system.

However, Armeria's `FileService` does not understand the path pattern (`[projectName]`), so it just tries to find `/app/products/my-project` file or `/app/products/my-project/index.html` from the file system which does not exist. 

I set `app/index.html` as a fallback and trigger re-routing in the client-side when the page is first loaded.

**Review by trustin — APPROVED** [review]
Thanks, @ikhoon !
